timeseries/src/autogluon/timeseries/models/autogluon_tabular/tabular_model.py

```diff
@@ -163,78 +140,130 @@ def apply_mask(
-
-            if self.metadata.past_covariates_real:
-                past_covariates_lags = get_lags(
-                    time_series[self.metadata.past_covariates_real], self._past_covariates_lag_indices
-                )
-                past_covariates_lags = apply_mask(
-                    past_covariates_lags,
-                    num_hidden=num_hidden,
-                    lag_indices=self._past_covariates_lag_indices,
-                    num_columns=len(self.metadata.past_covariates_real),
-                )
-                feature_dfs.append(past_covariates_lags)
+            num_rows = min(max_rows_per_item, len(ts))
+            features = np.full([num_rows, len(lag_indices)], fill_value=np.nan)
+            for i in range(1, num_rows + 1):
+                target_idx = len(ts) - i
+                selected_lags = lag_indices[lag_indices <= target_idx]
+                features[num_rows - i, np.arange(len(selected_lags))] = ts[target_idx - selected_lags]
+            if mask:
+                num_windows = (len(ts) - 1) // prediction_length
+                # We don't hide any past values for the first `remainder` values, otherwise the features will be all empty
+                remainder = len(ts) - num_windows * prediction_length
+                num_hidden = np.concatenate([np.zeros(remainder), np.tile(np.arange(prediction_length), num_windows)])
+                features = apply_mask(features, num_hidden[-num_rows:], lag_indices)
+            return features
 
-            if self.metadata.known_covariates_real:
-                known_covariates_lags = get_lags(
-                    time_series[self.metadata.known_covariates_real], self._known_covariates_lag_indices
+        def get_lag_features(
+            all_series: List[np.ndarray],
+            lag_indices: np.ndarray,
+            prediction_length: int,
+            max_rows_per_item: int,
+            mask: bool,
+            name: str,
+        ):
+            """Generate lag features for all time series in the dataset.
+
+            See the docstring of get_lags for the description of the parameters.
+            """
+            lags_per_item = Parallel(n_jobs=-1)(
+                delayed(get_lags)(
+                    ts,
+                    lag_indices,
+                    prediction_length=prediction_length,
+                    max_rows_per_item=max_rows_per_item,
+                    mask=mask,
                 )
-                feature_dfs.append(known_covariates_lags)
-
-            features = pd.concat(feature_dfs, axis=1)
-            return features
+                for ts in all_series
+            )
+            features = np.concatenate(lags_per_item)
+            return pd.DataFrame(features, columns=[f"{name}_lag_{idx}" for idx in lag_indices])
+
+        df = pd.DataFrame(data)
+        all_series = [ts for _, ts in df.droplevel(TIMESTAMP).groupby(level=ITEMID, sort=False)]
+
+        feature_dfs = []
+        for column_name in df.columns:
```

Comment:
[minor] we can simplify this logic into something like:
```python
column_name_to_mask_indices_mapping = {
  self.target: (True, self._target_lag_indices),
  self.metadata.past_covariates_real: (True, self._past_covariates_lag_indices),
  self.metadata.known_covariates_real:  (False, self._known_covariates_lag_indices)
}

mask, lag_indices = column_name_to_mask_indices_mapping. get(column_name, (None, None))
if mask is None: raise ValueError(f"Unexpected column {column_name} is not among target or covariates.")
```

Reply:
Not sure, I think this will be a bit more complicated since `self.metadata.past_covariates_real` and `self.metadata.known_covariates_real` are both lists, so we will need to have 3 loops - two when constructing `column_name_to_mask_indices_mapping` and one when iterating over the columns of the `df`.  

The current code takes more space then the suggestion, but I feel like the current version might be slightly easier to read because of the double for-loop in the suggestion.